Major CLI updates plus CLI data testing

ingydotnet opened the pull request.

ccoVeille commented on cmd/go-yaml/cli_test.go:
```diff
@@ -0,0 +1,193 @@
+
+	// Build the CLI binary if it doesn't exist
+	binaryPath := "../../go-yaml"
+	if _, err := os.Stat(binaryPath); os.IsNotExist(err) {
```
You should stop here and report an error if the file exists but is a folder.

Think about someone who created a go-yaml folder.

ccoVeille commented on cmd/go-yaml/event.go:
```diff
@@ -63,13 +66,27 @@ func ProcessEvents(profuse, compact, unmarshal bool) error {
 }
 
-// processEventsDecode uses Loader.Load for YAML processing
+// processEventsDecode uses yaml.Decoder for YAML processing with implicit field augmentation
```
This sounds a bad conflict resolution

Loader.Load is the one used now

ccoVeille commented on cmd/go-yaml/event.go:
```diff
@@ -491,11 +564,176 @@ func formatEventInfo(event *Event, profuse bool) *EventInfo {
+			info.Pos = fmt.Sprintf("%d/%d-%d", event.StartLine, event.StartColumn, event.EndColumn)
+		} else {
+			// Range across different lines
+			info.Pos = fmt.Sprintf("%d/%d-%d/%d", event.StartLine, event.StartColumn, event.EndLine, event.EndColumn)
```
The classic way to write it would be this


```suggestion
			info.Pos = fmt.Sprintf("%d:%d", event.StartLine, event.StartColumn)
		} else if event.StartLine == event.EndLine {
			// Range on same line
			info.Pos = fmt.Sprintf("%d:%d-%d", event.StartLine, event.StartColumn, event.EndColumn)
		} else {
			// Range across different lines
			info.Pos = fmt.Sprintf("%d:%d-%d:%d", event.StartLine, event.StartColumn, event.EndLine, event.EndColumn)
```

ingydotnet replied:
I made this change but it has the unfortunate side effect of quoting all the position values. That's why I chose the format that I did. Later though I think we can fix the encoder to not do that. It does that because of the YAML 1.1 rules. That's why I filed #231.

ccoVeille commented on cmd/go-yaml/token.go:
```diff
@@ -374,26 +375,47 @@ func formatTokenInfo(token *Token, profuse bool) *TokenInfo {
 		} else {
-			info.Pos = fmt.Sprintf("%d;%d-%d;%d", token.StartLine, token.StartColumn, token.EndLine, token.EndColumn)
+			// Range across different lines
+			info.Pos = fmt.Sprintf("%d/%d-%d/%d", token.StartLine, token.StartColumn, token.EndLine, token.EndColumn)
```
Same remark about the fact : should be used instead of /

ccoVeille commented on cmd/go-yaml/node.go:
```diff
@@ -87,24 +125,199 @@ func formatStyle(s yaml.Style) string {
+	case yaml.FlowStyle:
+		return "flow"
+	default:
+		return "plain"
```
I would use this.

```suggestion
	case 0:
		return "plain"
	default:
		return "unknown"
```

But maybe I'm too defensive.

But formatStyle returns ""

ingydotnet commented on cmd/go-yaml/event.go:
```diff
@@ -78,252 +95,307 @@ func processEventsDecode(profuse, compact bool) error {
+	}
+
 	// Split input into documents
 	documents := bytes.Split(input, []byte("---"))
```
Not a bug:
```
$ ./go-yaml <<<'
> a: "---"
b: "---"
> ' -e
- {event: DOCUMENT-START}
- {event: MAPPING-START}
- {event: SCALAR, value: a}
- {event: SCALAR, value: '---', style: Double}
- {event: SCALAR, value: b}
- {event: SCALAR, value: '---', style: Double}
- {event: MAPPING-END}
- {event: DOCUMENT-END}
```

ingydotnet commented:
I'm putting this back into draft mode because I have some newer and bigger plans for how this should work.

ingydotnet commented:
@ccoVeille I think you'll like 6108ac5 because I replaced the few options I added with a generic `-o` `--option` flag to specify any possible combination of options!

ccoVeille commented on cmd/go-yaml/cli_test.go:
```diff
@@ -0,0 +1,223 @@
+		}
+		if i < len(actLines) {
+			actLine = actLines[i]
+		}
```
Nitpicking: You reduced maxLine to the lowest of both slice, so here you can use that


```suggestion
		expLine: = expLines[i]
		actLine: = actLines[i]
```

ingydotnet replied:
This sets maxLines to the maximum (not minimum) of both lengths. This is intentional - the diff needs to show trailing lines when arrays have different lengths. The bounds checks are required because we iterate up to the longer array.

ccoVeille commented on cmd/go-yaml/cli_test.go:
```diff
@@ -0,0 +1,223 @@
+		}
+
+		if expLine != actLine {
+			result.WriteString(fmt.Sprintf("Line %d:\n", i+1))
```
This is usually written like this


```suggestion
			fmt.Fsprintf(result, "Line %d:\n", i+1)
```

ccoVeille commented on cmd/go-yaml/json.go:
```diff
@@ -15,17 +15,21 @@ import (
-	if unmarshal {
-		return processJSONUnmarshal(pretty)
+// ProcessJSON reads YAML from reader and outputs JSON encoding
+func ProcessJSON(reader io.Reader, pretty, unmarshalMode, decodeMode bool, opts []yaml.Option) error {
```
Usually, we are using variadic


```suggestion
func ProcessJSON(reader io.Reader, pretty, unmarshalMode, decodeMode bool, opts ...yaml.Option) error {
```

ccoVeille commented on cmd/go-yaml/json.go:
```diff
@@ -15,17 +15,21 @@ import (
-func processJSONDecode(pretty bool) error {
-	loader, err := yaml.NewLoader(os.Stdin)
+// processJSONLoad uses Loader.Load for YAML processing with options
+func processJSONLoad(reader io.Reader, pretty bool, opts []yaml.Option) error {
```
Same here


```suggestion
func processJSONLoad(reader io.Reader, pretty bool, opts ...yaml.Option) error {
```

ccoVeille commented on cmd/go-yaml/main.go:
```diff
@@ -15,12 +15,278 @@ import (
+	if s == nil {
+		return ""
+	}
+	return fmt.Sprintf("%v", *s)
```
No formatting, this should be preferred

```suggestion
	return fmt.Sprint(*s)
```

ccoVeille replied:
I'll add linters when the PR paces will slow down

ingydotnet replied:
99481be

ccoVeille reviewed (approved):
The PR became wild with huge amount of code changes.

It's OK, it's a CLI used for debug purpose.

I'm a bit worried the CLI become as big as the library itself. It's uneasy to maintain.

But these changes are interesting and needed.

Some remarks anyway. But they can be ignored if there is a need to merge quickly.

ccoVeille commented on cmd/go-yaml/cli_test.go:
```diff
@@ -0,0 +1,334 @@
+		fmt.Fprintf(os.Stderr, "Failed to create temp dir: %v\n", err)
+		os.Exit(1)
+	}
+	defer os.RemoveAll(tmpDir)
```
t.TempDir could be used and you could simplify here

ingydotnet replied:
ca2ee8b

ccoVeille commented on cmd/go-yaml/cli_test.go:
```diff
@@ -0,0 +1,334 @@
+// TestCLICommands runs command-based tests from testdata/cmd/*.yaml
+func TestCLICommands(t *testing.T) {
+	// Find all test files in testdata/cmd/
+	testFiles, err := filepath.Glob("testdata/cmd/*.yaml")
```
This should be the way to be architecture agnostic

```suggestion
	testFiles, err := filepath.Glob(filepath.Join("testdata", "cmd", "*.yaml"))
```

ingydotnet replied:
e3c138e

ccoVeille commented on cmd/go-yaml/main.go:
```diff
@@ -15,27 +15,324 @@ import (
+		opts = append(opts, opt...)
+	}
+	return opts, nil
+}
```
This doesn't make sense to me


```suggestion
func parseOptionFlags(s string) ([]yaml.Option, error) {
	parts := strings.Split(s, ",")
	var opts []yaml.Option
	for _, part := range parts {
		trimmed := strings.TrimSpace(part)
		if trimmed == "" {
			continue
		}
		opt, err := parseOneOption(trimmed)
		if err != nil {
			return nil, err
		}
		opts = append(opts, opt...)
	}
	return opts, nil
}
```

Am I missing something?

ingydotnet commented on cmd/go-yaml/json.go:
```diff
@@ -54,12 +58,40 @@ func processJSONDecode(pretty bool) error {
 }
 
+// processJSONDecode uses deprecated Decoder.Decode for YAML processing (no options support)
+func processJSONDecode(reader io.Reader, pretty bool, opts []yaml.Option) error {
```
8e514fd

ingydotnet commented:
I'm going to merge this now, without requiring a full review from 2 team members. 
I feel confident in the tests and behavior. 
And it's only a debugging CLI. The diff is huge:
```
li-fixes-and-refactoring $ git diff origin/main | wc -l
5421
```
That's a lot to expect people to review. 
This should be the last major refactor of this size; at least for a while. 
I want to use this code to help me with the plugin API. 

I discussed this with @colinjlacy in Slack and he was ok with me merging if I was. And I am.
